DFE-Digital/sts-plan-technology-for-your-school
# Feat: Add validation to contentful fields

## Overview

Addresses ticket [#234864](https://dfe-ssp.visualstudio.com/s190-schools-technology-services/_workitems/edit/234864)

Introduces validation on most text fields within contentful to disallow leading or trailing whitespace.
For slugs it disallows any whitespace.
For questions it disallows leading whitespaces and requires terminating with a question mark.

## Changes

- Adds a contentful management script that can be used to delete all links to content that no longer exists
  - This was made a bit difficult by the maximum of 1000 entries brought back by the api
  - so it is done by going through each content type in turn as each one has below that limit
- Adds a contentful migration script that goes through each content type and
  - For any slug fields, strips whitespace, and adds no whitespace validation
  - For any text fields, strips whitespace and adds no leading/trailing whitespace validation
  - For questions, strips whitespace and adds no leading whitespace/must end in ? validation

Other minor changes
- Changes the validate environment from the management project to retrieve valid environments programatically so that we're protected should they change in future
- moves the content-types iteration into helpers

## How to review the PR

I've already run both scripts on dev but I recommend running both locally

To test the management script
- Add some new linked content to a few entries, and then delete the new content (e.g. create a new answer, add it to an existing question, then delete the answer so that the question has an answer that is 'missing or inaccessible')
- Check the management script removes the bad reference
- Check that it does not affect any of the good references or any other fields

To test the migration script
- Go into the content types and remove the regex validation from a content type
- Add some leading and trailing whitespace to various text fields of the type you updated
- Run the migration script and check that the whitespace is removed and the validation is re-added

## Release requirements

We will need to run the management script on staging to remove any old references, followed by the migration script

## Checklist
